File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
def solution(bridge_length, weight, truck_weights):
    ing = []
    cur_weight = 0
    time = 0

    while truck_weights or ing:
        time += 1
        
        if len(ing) >= bridge_length:
            cur_weight -= ing.pop(0)

        if truck_weights and truck_weights[0] + cur_weight <= weight: 
            ing.append(truck_weights.pop(0))
            cur_weight += ing[-1]
        else:
            ing.append(0)
            
        if not truck_weights and max(ing) == 0:
            break
    
    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<Integer> bridge = new LinkedList<>();
        int time = 0;
        int curWeight = 0;
        int done = 0;
        int idx = 0;
        
        
        while (done < truck_weights.length) {
            if (idx < truck_weights.length && curWeight + truck_weights[idx] <= weight) {
                curWeight += truck_weights[idx];
                bridge.add(truck_weights[idx++]);
            }
            else {
                bridge.add(0);
            }
            
            if (bridge.size() == bridge_length) {
                int weightToRemove = bridge.remove();
                if (weightToRemove > 0) {
                    done++;
                    curWeight -= weightToRemove;
                }
            }
            time++;
        }
        return time + 1;
    }
}